### ISSUE TITLE
Amiberry - virtual keyboard

### ISSUE THREAD

**Issue by tukicat** [BlitterStudio/amiberry#224]
I have complied the "dev" version and I am doing a little testing with my my rpi3 and IPAC so I can provide a couple of Retroarch.cfg for IPAC users for your wiki.

I just wondered.......does Amiberry have a an option to display a virtual keyboard on screen?

**Comment by midwan** [comment]
@tukicat 
Not yet, but it's planned to be added in the near future. We already have sample code for it, but we haven't had time to implement it in Amiberry yet.

**Comment by tukicat** [comment]
Great thanks for the info. A while ago I helped Chui with virtual keyboard gfx for the UAE4ALL port for the Dingoo A320, I would be happy to help if there is anything you need for Amiberry.

**Comment by midwan** [comment]
@tukicat 
Fantastic, help in the project is always welcome! :)
For now, the main focus is get things in order for the next stable release, so we can finally merge "dev" back into "master". Once that is done, we can start adding new features into "dev" again, so this will be one of them.

Let's keep this issue for future reference, and revisit it after the new version is out.

**Comment by tukicat** [comment]
No problem, happy to help when you are ready.

**Comment by rsn8887** [comment]
ScHlAuChi also made some great keyboard images I used in UAE4All2 for Vita. If you are interested, here are the images:
https://drive.google.com/open?id=1W9GCWUZl_cDjMFsUnLja5VTaNZLKPMv1

Here is an example of one of the keyboard images in use:
![Screenshot of UAE4All2 Vita_mod](https://s13.postimg.org/9r9hi62mf/uae4all-vita.jpg)

**Comment by ZappaUtopia** [comment]
That keyboard looks awesome! Hopefully it can be used (and with a bit of transparency so you can see what's going on). I also hope this can replace the Android one. Good job.

**Comment by rsn8887** [comment]
The archive link I posted also includes a version with Shift depressed (capital letters etc) and versions for uk and German keymaps :)

**Comment by HoraceAndTheSpider** [comment]
we do already have a gfx in mind, but maybe it is something that could have several options (if the layout can be tabled in some way) 

www.djcresswell.com/amigaonline/AmigaKeyboardPSP.png

**Comment by ZappaUtopia** [comment]
I think @rsn8887 's would be better IMHO as it's clearer and can still be seen clearly with some transparency which will be vital when selecting keys as you can see the image underneath. That cropped photo of the Amiga keyboard looks curved at the left and right sides, you could have at least used a straight picture :)

**Comment by HoraceAndTheSpider** [comment]
The transparency would be added by the graphics drawing routine - should be no need to have it on the picture itself. (otherwise, what would you do, dither every other pixel?)

Used this one for years on another popular Amiga emulator, so no problem with the curve.  
Not being rude, but I really dont think the other one looks very Amiga. It has more of an 8-bit feel.

As i said, it may be possible to have multiple options anyway.

**Comment by rsn8887** [comment]
The 8bit feel of ScHlAuChi's keyboards I uploaded came from necessity: The image was made to fit into a 320*200 minimum screen resolution (the smaller the screen resolution, the faster the emulator).

**Comment by tukicat** [comment]
Hi, I am glad that there are some options for virtual keyboards on Amiberry. As I say a while ago I did the virtual keyboard for Chui's Dingoo A320 port of UAE4ALL, I believe it may also have been used for the Pandora port. 

Completely by chance I was at friends and noticed he had an Armiga and the virtual keyboard on this is the same one I produced for the A320. 

The virtual keyboard has a keyboard layout and when you hover the cursor over each key it shows and enlarged version to the side. However, I could make better version using the enlarged keys.

If you did want to use this, and I am by no way saying it is the best option, then I probably just need check with Chui that it is ok, as I produced it for his use some time ago. 

![image](https://user-images.githubusercontent.com/35096478/35880948-2f4f1a82-0b77-11e8-9c39-5a3772da4019.png)

**Comment by CypherXG** [comment]
![specialkeyboard](https://user-images.githubusercontent.com/7895130/37239468-fd47a746-243c-11e8-949c-9458e4f40210.png)
Here you can have my iAmiga Keyboard that i use for my iPad Amiga Emu. I don't need this anymore.

**Comment by Darknior** [comment]
WOAW i love it, please use this one <3

**Comment by CypherXG** [comment]
:-) Thx

**Comment by Darknior** [comment]
All keys are square, it will easier to use :)
It should be cool to use a light blue on KEY OVER and a dark blue on KEY PRESS, and possibly leave the dark blue on a key that remains active as the key SHIFT or SHIFT for example. It's just a idea already use on the AMIGA emulator port on xbox.

**Comment by waal** [comment]
@CypherXG 

Lovely, I believe that should be the one ! Good work, man.

**Comment by Gilou9999** [comment]
That's what i call a nice virtual keyboard! It will brings a lot to the experience, playing our old games with a nice one like this!

**Comment by midwan** [comment]
As a first step, I've added the `vbkd` files from @rsn8887's UAE4All port in the repo (as well as other PSP2 related stuff, with the aim of making Amiberry work on PSVita/PSTV later).

**Comment by rsn8887** [comment]
Note the files have since been updated (now with French keyboard, too): 
https://github.com/rsn8887/uae4all2/tree/master/src/vkbd

**Comment by midwan** [comment]
@rsn8887 
Thanks for the heads up! In a funny coincidence, I was experimenting with compiling Amiberry for the PSP2/PSVita/PSTV earlier today. :)

I had to remove the vbkd for now, since it's still using SDL1 calls and we've moved to SDL2 (I noticed that sdl2 to sdl1 header). I'll merge the latest you have and convert them to pure SDL2 at some later point, as I need to get other things fixed with it first (I got several files compiled normally, until I got to the `fpp_native` file - which requires `fenv.h` to get some definitions from there, but those could not be found in the vitasdk environment. Anyway, that's a different issue than this one...

**Comment by sh0wst0pper** [comment]
Sorry to revive an old thread but is there any progress on a virtual keyboard for Amiberry on retropie? I have this emulator on my switch (which has a virtual keyboard) and have just built a bartop arcade and this was the 1st emulator i had to have running. Most games dont require it but it would be nice to have.

If not then no problem and many thanks for this amazing app!

**Comment by davidgrohl** [comment]
any news about virtual keyboard?

**Comment by midwan** [comment]
@davidgrohl 
Yes, coming soon... :)

**Comment by davidgrohl** [comment]
Just updated to version 5.1. can you confirm no virtual keyboard added yet?

**Comment by midwan** [comment]
> Just updated to version 5.1. can you confirm no virtual keyboard added yet?

It's not part of 5.1, as it wasn't quite ready to be included there. It still needs some work, so it will come with a future update.

**Comment by Tobias2K18** [comment]
... any news?

**Comment by erikbernsen** [comment]
I was planning on adding this myself until I see it’s already being worked on. Is there somewhere I can try this out or contribute?

**Comment by midwan** [comment]
The `dev` branch has some work-in-progress implementation, but I haven't had time to revisit this lately. 
The current status is that the virtual keyboard is brought up on the screeen (with transparency) on-demand, and hidden again, but you can't select any keys to press yet. There are multiple themes for keyboards available as well, found in `data/vkbd` - the plan was to choose the one you prefer, or even use your own.

**Comment by erikbernsen** [comment]
Thanks. I finally got time to start looking into this. I downloaded the dev branch and compiled it. But didn't see that there is any functionality to bring up the virtual keyboard. The function vkdb_init and vkbd_redraw seem implemented but as far as I can see nowhere used yet. Or I'm overlooking it. As a very small hack I call vkbd_init and vkbd_redraw somewhere myself and indeed see the keyboard. It's a start to understand the code.

<img width="774" alt="Screenshot" src="https://user-images.githubusercontent.com/3802779/211216814-e39f47d1-9744-4036-bca9-0879cf64fdae.png">

I have some questions; if this is addressed somewhere in documentation, could you please direct me to it.
- Is there a recommended development environment? 
- When I run cmake on the dev branch without modifying CMakeLists.txt I get errors:

CMake Error at CMakeLists.txt:21 (add_executable):
  Cannot find source file:

    src/machdep/m68k.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc


CMake Error at CMakeLists.txt:21 (add_executable):
  No SOURCES given to target: amiberry

Indeed     src/machdep/m68k.cpp doesn't exist on disk. I'm just calling cmake as 

cmake . 

from the Amiberry directory. Am I doing something wrong here? I'm not very familiar with cmake.

**Comment by midwan** [comment]
@erikbernsen 
Hm, I had added a custom event which allowed you to trigger the on-screen keyboard with a press of a button (e.g. from a controller), but I can't find that code in the branch. Perhaps I had it in a separate branch which got deleted over time, I'm not sure at this point... :(

CMake works on the `x86` branch, and I made some merges from there back to `master` and `dev`, but it appears I missed that file. Depending on the platform you're working on, you can use different options:
- VSCode (multi-platform and configurable)
- Visual Studio (Windows only, compiles remotely on a device using the solution included or locally using WSL2)
- CLion (multi-platform, compiles remotely on a device)

I am usually on Windows, so I mostly use either Visual Studio or VSCode, depending on what I'm doing.
Let's take the rest on a new Discussion topic however, to avoid having such details on this issue: https://github.com/BlitterStudio/amiberry/discussions

**Comment by erikbernsen** [comment]
Thanks for confirming that. Could you let me know if you still have these changes locally as it would help. I was thinking a custom event was the way to enable this indeed.

**Comment by midwan** [comment]
> Thanks for confirming that. Could you let me know if you still have these changes locally as it would help. I was thinking a custom event was the way to enable this indeed.

Unfortunately it seems that I lost those, when I deleted that branch (I was sure the changes were merged!).
But they are not difficult to re-implement, and in fact I should sync that vkbd stuff across branches as well.

**Comment by erikbernsen** [comment]
Thanks for that. I continued working on this today and made quite some progress. I got it to the point that you can toggle the virtual keyboard on or off; use the joystick to select a key; press fire to type the key. A lot of stuff is still missing like shift, sticky keys, nice indication what's the currently selected key; position of the keyboard; selecting different language keyboard; select translucency of keyboard.

**Comment by midwan** [comment]
@erikbernsen Sounds good, I'm happy you're making progress with it! :)

**Comment by erikbernsen** [comment]
@midwan I'm almost done with making it work for my setup and to my liking. I can only test for my setup and didn't try yet to make it work for other controllers and hardware.  Will try to figure out how to create PR so you can review the code and maybe more people can test. Should I create a PR into dev branch or what's the best approach?

**Comment by midwan** [comment]
@erikbernsen 
That's great news! Please submit a PR towards `master` if you have something that works, and we'll test it from there.
